yxi/xml/xml_generator.py

```diff
@@ -0,0 +1,113 @@
+for the spec this is based on.
+"""
+size_key = "total_size"
+width_key = "data_width"
+
+def gen_xml(yxi):
+  mems = yxi["memories"]
+  check_mems_wellformed(mems)
+
+  root = Element("root", {"versionMajor": "1", "versionMinor": "6"})
+  kernel = SubElement(root, "kernel", {
+    "name" : yxi["toplevel"],
+    "language": "ip_c",
+    # TODO: Make sure this matches component.xml, Namely the `Toplevel` part.
+    #  See https://docs.amd.com/r/en-US/ug1393-vitis-application-acceleration/RTL-Kernel-XML-File
+    "vlnv": "capra.cs.cornell.edu:kernel:Toplevel:1.0",
+    "attributes": "",
+    "preferredWorkGroupSizeMultiple" : "0",
+    "workGroupSize" : "1",
+    "hwControlProtocol": "ap_ctrl_hs"
+  })
+
+  # Construct ports
+  ports = SubElement(kernel, "ports")
+  # The subordinates XRT - AXI controller is added outside of the programs memory interface.
+  SubElement(ports, "port", {
+    "name": "S_AXI_CONTROL",
+    "mode": "slave",
+    "range": "0x1000", # XXX: Why is this 0x1000? Taken from the Xilinx examples  Maybe could be dynamically calculated?
+    "dataWidth" : "32",
+    "portType" : "addressable",
+    "base" : "0x0"
+  })
+
+  for mem in mems:
+    SubElement(ports, "port", {
+      "name": f"m_axi_{mem['name']}",
+      "mode": "master",
+      #NOTE(nathaniel): In the Xilinx examples range is usually 0xFFFFFF... but this should be fine for us?
+      "range": f"{hex(size_in_bytes(mem))}",
+      "dataWidth": f"{mem[width_key]}",
+      "portType": "addressable",
+      "base": "0x0"
+    })
+
+  # Construct Args
+  args = SubElement(kernel, "args")
+  # XRT spec starts args addresses at 0x10
+  args_addr = 0x10
+  for i, mem in enumerate(mems):
+    SubElement(args, "arg", {
+      "name": f"{mem['name']}",
+      "addressQualifier": "1", # 1 denotes the arguments as a global memory, 
+      "id" : f"{i}",
+      "port" : f"m_axi_{mem['name']}",
+      "size" : f"0x8", # XRT expects AXI manager interfaces that are 64 bits wide 
+      "offset" : f"{hex(args_addr + (i * 8))}",
+      "type" : "int*", # TODO: Old xml generator defaults to `int*`, I imagine in practice we want to get this from our .dat
+      "hostOffset" : "0x0",
+      "hostSize" : "0x8", # Seems to be the same as `size`, unclear how they differ
+    })
+
+  
+  
+
+  return root
+
+def size_in_bytes(mem):
+  return mem[size_key] * mem[width_key] // 8
+
+#TODO: Import from axi_generator instead of copy pasting here
+def check_mems_wellformed(mems):
+  """Checks if memories from yxi are well formed. Returns true if they are, false otherwise."""
+  for mem in mems:
+      assert (
+          mem[width_key] % 8 == 0
+      ), "Width must be a multiple of 8 to alow byte addressing to host"
+      assert log2(
+          mem[width_key]
+      ).is_integer(), "Width must be a power of 2 to be correctly described by xSIZE"
+      assert mem[size_key] > 0, "Memory size must be greater than 0"
+
+def prettify(elem):
```

Comment:
Seems to be unused?

Reply:
This is mainly useful for debugging. I'll add a comment, but can also delete if we don't foresee needing this.